File: backend/src/core/orchestrator.ts
import type { MissionDefinition, MissionState } from "../types";
import type { ResolvedStack } from "../capabilities";
import type { RemediationAction } from "../capabilities/remediation";
import { createInitialAgents } from "./state-helpers";
import { getMissionPlan } from "../missions";
import { getRunStore } from "./run-store";

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export function createInitialMissionState(
  mission: MissionDefinition
): MissionState {
  return {
    id: mission.id,
    status: "idle",
    agents: createInitialAgents(mission.agents),
    timeline: [],
    toolCalls: [],
    evidence: [],
    cards: [],
    confidence: 0,
  };
}

function extractPendingAction(state: MissionState): RemediationAction | undefined {
  const approval = [...state.cards]
    .reverse()
    .find((c) => c.type === "approval");
  if (!approval) return undefined;
  const d = approval.data;
  return {
    action: String(d.action ?? ""),
    from: d.from as string | undefined,
    to: d.to as string | undefined,
    namespace: d.namespace as string | undefined,
    nodePool: d.nodePool as string | undefined,
    expectedImprovement: d.expectedImprovement as string | undefined,
    risk: d.risk as string | undefined,
    cost: d.cost as string | undefined,
    requiresApproval: Boolean(d.requiresApproval),
  };
}

/**
 * MissionOrchestrator — runs capability-driven mission plans and streams
 * MissionState snapshots through the run store (SSE consumers).
 */
export class MissionOrchestrator {
  async start(params: {
    runId: string;
    mission: MissionDefinition;
    stack: ResolvedStack;
  }): Promise<void> {
    const store = getRunStore();
    const plan = await getMissionPlan(params.mission.id);
    let state = createInitialMissionState(params.mission);
    state = { ...state, status: "running" };
    store.updateState(params.runId, state);

    for (const step of plan.steps) {
      const run = store.get(params.runId);
      if (!run || run.abort) return;

      await sleep(step.delay);
      state = await step.run({
        state,
        stack: params.stack,
        mission: params.mission,
      });
      store.updateState(params.runId, { ...state });

      if (state.status === "awaiting-approval") {
        store.setPendingAction(params.runId, extractPendingAction(state));
        return;
      }
    }

    if (state.status !== "awaiting-approval" && state.status !== "complete") {
      state = { ...state, status: "complete" };
      store.updateState(params.runId, state);
    }
  }

  async approve(runId: string, stack: ResolvedStack): Promise<MissionState> {
    const store = getRunStore();
    const run = store.get(runId);
    if (!run) throw new Error(`Unknown run: ${runId}`);

    const action = run.pendingAction ?? {
      action: "Approved remediation",
    };
    const result = await stack.remediation.apply(action);
    const state: MissionState = {
      ...run.state,
      status: "complete",
      timeline: [
        ...run.state.timeline,
        {
          id: `tl-${run.state.timeline.length + 1}`,
          timestamp: new Date().toLocaleTimeString("en-US", { hour12: false }),
          agentId: "policy",
          message: result.message,
          type: "recommendation",
        },
      ],
    };
    store.updateState(runId, state);
    store.setPendingAction(runId, undefined);
    return state;
  }

  async reject(runId: string, stack: ResolvedStack): Promise<MissionState> {
    const store = getRunStore();
    const run = store.get(runId);
    if (!run) throw new Error(`Unknown run: ${runId}`);

    const action = run.pendingAction ?? { action: "Rejected remediation" };
    const result = await stack.remediation.reject(action);
    const state: MissionState = {
      ...run.state,
      status: "complete",
      timeline: [
        ...run.state.timeline,
        {
          id: `tl-${run.state.timeline.length + 1}`,
          timestamp: new Date().toLocaleTimeString("en-US", { hour12: false }),
          agentId: "policy",
          message: result.message,
          type: "observation",
        },
      ],
    };
    store.updateState(runId, state);
    store.setPendingAction(runId, undefined);
    return state;
  }
}

export function getOrchestrator(): MissionOrchestrator {
  return new MissionOrchestrator();
}

File: backend/kip/core/orchestrator.py
from __future__ import annotations

import asyncio

from kip.capabilities import RemediationAction, ResolvedStack
from kip.core.run_store import get_run_store
from kip.core.state_helpers import create_initial_agents, now_timestamp
from kip.missions import get_mission_plan
from kip.types import MissionDefinition, MissionState, TimelineEvent


def create_initial_mission_state(mission: MissionDefinition) -> MissionState:
    return MissionState(
        id=mission.id,
        status="idle",
        agents=create_initial_agents(mission.agents),
        timeline=[],
        toolCalls=[],
        evidence=[],
        cards=[],
        confidence=0,
    )


def _extract_pending_action(state: MissionState) -> RemediationAction | None:
    approval = next(
        (c for c in reversed(state.cards) if c.type == "approval"),
        None,
    )
    if not approval:
        return None
    d = approval.data
    return RemediationAction(
        action=str(d.get("action") or ""),
        from_=d.get("from"),
        to=d.get("to"),
        namespace=d.get("namespace"),
        nodePool=d.get("nodePool"),
        expectedImprovement=d.get("expectedImprovement"),
        risk=d.get("risk"),
        cost=d.get("cost"),
        requiresApproval=bool(d.get("requiresApproval")),
    )


class MissionOrchestrator:
    async def start(
        self,
        *,
        run_id: str,
        mission: MissionDefinition,
        stack: ResolvedStack,
    ) -> None:
        store = get_run_store()
        plan = get_mission_plan(mission.id)
        state = create_initial_mission_state(mission)
        state = state.model_copy(update={"status": "running"})
        store.update_state(run_id, state)

        for step in plan.steps:
            run = store.get(run_id)
            if not run or run.abort:
                return

            await asyncio.sleep(step.delay)
            state = await step.run(state, stack, mission)
            store.update_state(run_id, state)

            if state.status == "awaiting-approval":
                store.set_pending_action(run_id, _extract_pending_action(state))
                return

        if state.status not in ("awaiting-approval", "complete"):
            state = state.model_copy(update={"status": "complete"})
            store.update_state(run_id, state)

    async def approve(self, run_id: str, stack: ResolvedStack) -> MissionState:
        store = get_run_store()
        run = store.get(run_id)
        if not run:
            raise ValueError(f"Unknown run: {run_id}")

        action = run.pending_action or RemediationAction(action="Approved remediation")
        result = await stack.remediation.apply(action)
        timeline = list(run.state.timeline) + [
            TimelineEvent(
                id=f"tl-{len(run.state.timeline) + 1}",
                timestamp=now_timestamp(),
                agentId="policy",
                message=result.message,
                type="recommendation",
            )
        ]
        state = run.state.model_copy(
            update={"status": "complete", "timeline": timeline}
        )
        store.update_state(run_id, state)
        store.set_pending_action(run_id, None)
        return state

    async def reject(self, run_id: str, stack: ResolvedStack) -> MissionState:
        store = get_run_store()
        run = store.get(run_id)
        if not run:
            raise ValueError(f"Unknown run: {run_id}")

        action = run.pending_action or RemediationAction(action="Rejected remediation")
        result = await stack.remediation.reject(action)
        timeline = list(run.state.timeline) + [
            TimelineEvent(
                id=f"tl-{len(run.state.timeline) + 1}",
                timestamp=now_timestamp(),
                agentId="policy",
                message=result.message,
                type="observation",
            )
        ]
        state = run.state.model_copy(
            update={"status": "complete", "timeline": timeline}
        )
        store.update_state(run_id, state)
        store.set_pending_action(run_id, None)
        return state


def get_orchestrator() -> MissionOrchestrator:
    return MissionOrchestrator()
